### ISSUE TITLE
ArgumentTransform Attribute

### ISSUE THREAD

**Issue by StartAutomating** [PowerShell/PowerShell#20112]
### Summary of the new feature / enhancement

There is a base type for a argument transformations, but there are not too many implementations of this attribute that are public.

Having a generic ArgumentTransform attribute would be highly useful for a variety of scenarios.

On practical example of this is the solution to #19704 (pr #19705).

In this scenario, we wanted to add more granularity to Wait-Event and the PowerShell committee wished to provide a TimeSpan or double value for total seconds.

Other scenarios include:

* Converting various forms of Color to a hex value
* Binding to a known property of a complex object
* Extracting arguments from nested properties of objects

### Proposed technical implementation details (optional)

A relatively simple ArgumentTransform attribute should exist, with two properties:

* PropertyName (the name of the property on an input object that should serve as the value)
* TransformScript (a [ScriptBlock] that can convert the value arbitrarily)

**Comment by Jaykul** [comment]
Why would it need a PropertyName value? Wouldn't you just hard code that into the script?

**Comment by StartAutomating** [comment]
> Why would it need a PropertyName value? Wouldn't you just hard code that into the script?

You're right.  Removed the PropertyName and simplified the implementation quite a bit.

**Comment by vexx32** [comment]
WG-Engine discussed this one today, we agree on including the currently proposed transformation attribute. We are unsure whether the originally proposed PropertyName is _necessary_ although it was raised that the symmetry with things like `ValueFromPipelineByPropertyName` is a potentially interesting prospect to consider.

**Comment by StartAutomating** [comment]
Commenting to keep issue open.  Will loop back on associated prs.
